### PR TITLE
Add Move Token Deployment Script

### PR DESCRIPTION
**Pull request by 0xPrimata**
# Summary
- **RFCs**: $\emptyset$.
- **Categories**:  `scripts`.

Adds a script to deploy Move Token

# Changelog

* DeployMoveToken.s.sol
* DeployMoveTokenMultisig.s.sol
* Moves _mint() functionality to the deployment contract (MOVEToken.sol)
* Entire supply is minted to designated multisig
* Changes inheritance to ERC20Upgradeable for a state of art initial deployment.
* Legacy MOVEToken.sol -> MOVETokenV2.sol as we prep and propose what the next version of the token should look like.
* Tests handle Upgrade from MOVEToken to MOVETokenV2
* Tests cover proposer key compromised

# Testing

All tests are passing

# Outstanding issues

We might want to rename some variables like "multisig" to "foundation"

DeployMoveTokenMultisig.s.sol is currently broken

### CODE REVIEW

**Comment by l-monninger** [comment]
We're going to want to setup the `Timelock` to be controlled by a multisig, per governance roadmap. We'll talk about this later today.

File: protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        signers[3] = vm.addr(3);
+        signers[4] = vm.addr(4);
+
+        // DEPLOYMENT USING SAFE PROXY FACTORY
+        safeProxyFactory = SafeProxyFactory(0x4e1DCf7AD4e460CfD30791CCC4F9c8a4f820ec67);
+        safeAddress = payable(
+            address(
+                safeProxyFactory.createProxyWithNonce(
+                    safeSingleton,
+                    abi.encodeWithSignature(safeSetupSignature, signers, threshold, zero, "0x", 0xfd0732Dc9E303f09fCEf3a7388Ad10A83459Ec99, zero, 0, payable(zero)),
```
Comment (l-monninger) [inline protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:60]: Comment on address.

File: protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        );
+        safe = Safe(safeAddress);
+        // DEPLOYMENT USING SAFE PROXY FACTORY
+
+        // DEPLOYMENT USING SAFE PROXY
+        // safe = Safe(payable(address(new SafeProxy(safeSingleton))));
+        // safe.setup(signers, 3, zero, "0x", 0xfd0732Dc9E303f09fCEf3a7388Ad10A83459Ec99, zero, 0, payable(zero));
+        // DEPLOYMENT USING SAFE PROXY
+
+        uint256 minDelay = 1 days;
```
Comment (l-monninger) [inline protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:73]: I thought we were doing 48 hours?

File: protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        (uint8 v2, bytes32 r2, bytes32 s2) = vm.sign(2, digest);
+        (uint8 v3, bytes32 r3, bytes32 s3) = vm.sign(3, digest);
+
+        console.log("v1: ", v1);
+        console.log("v2: ", v2);
+        console.log("v3: ", v3);
+        signatures = abi.encodePacked(r1, s1, v1, r2, s2, v2);
+    }
+
+    function run() external {
```
Comment (l-monninger) [inline protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:39]: This needs to be chunked--either with sections clearly delineated by comments or with helper methods.

**Review by l-monninger — CHANGES_REQUESTED** [review]
Please provide docs describing intended usage of scripts.

File: protocol-units/settlement/mcr/contracts/test/token/MOVEToken.t.sol [hunk 4]
```diff
@@ -2,51 +2,285 @@
...
+    function testGrants() public {
+        testUpgradeFromTimelock();
+
+        // Check the token details
+        assertEq(MOVETokenV2(address(token)).hasRole(MOVETokenV2(address(token)).MINTER_ROLE(), multisig), true);
+    }
+
+    function testMint() public {
+        testUpgradeFromTimelock();
+        uint256 intialBalance = MOVETokenV2(address(token)).balanceOf(address(0x1337));
```
[the remark this answers is not in the document]
Comment (0xPrimata) [reply protocol-units/settlement/mcr/contracts/test/token/MOVEToken.t.sol:181]: testUpgradeFromTimelock(); is ran and the token proxy address is now using MOVETokenV2 implementation. To enforce it's interface, we use the MOVETokenV2 contract as the interface.

File: protocol-units/settlement/mcr/contracts/script/DeployMoveToken.s.sol [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    address moveAdmin;
+    bytes32 public salt = 0xc000000000000000000000002774b8b4881d594b03ff8a93f4cad69407c90350;
+    bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+
+    function run() external {
+        uint256 signer = vm.envUint("PRIVATE_KEY");
+        vm.startBroadcast(signer);
+
+        // forge script DeployMoveToken --fork-url https://eth-sepolia.api.onfinality.io/public
+        safe = Safe(payable(address(0x00db70A9e12537495C359581b7b3Bc3a69379A00)));
```
[the remark this answers is not in the document]
Comment (0xPrimata) [reply protocol-units/settlement/mcr/contracts/script/DeployMoveToken.s.sol:41]: okay

File: protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol (hunk 3 above)
Comment (0xPrimata) [reply protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:39]: Not sure if I understood this one. What needs to be chunked?

Comment (0xPrimata) [reply protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:39]: dropped this bellow as it's just utility for this specific contract. No need to move it elsewhere.

Comment (l-monninger) [reply protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:29]: This run method is too verbose. It should be broken into smaller helper methods.

File: protocol-units/settlement/mcr/contracts/test/token/MOVEToken.t.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (l-monninger) [reply protocol-units/settlement/mcr/contracts/test/token/MOVEToken.t.sol:181]: But, why is that test not in the `MOVEV2Token.t.sol`?

File: protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol (hunk 3 above)
Comment (0xPrimata) [reply protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:29]: already cleared.

File: protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        uint256 signer = vm.envUint("PRIVATE_KEY");
+        vm.startBroadcast(signer);
+        MOVEToken moveImplementation = new MOVEToken();
+
+        // forge script DeployMoveTokenMultisig --fork-url https://eth-sepolia.api.onfinality.io/public
+
+        address[] memory signers = new address[](5);
+
+        signers[0] = vm.addr(signer);
+        signers[1] = vm.addr(1);
```
[the remark this answers is not in the document]
Comment (l-monninger) [reply protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:39]: @0xPrimata The signers are still not explained. Leave a comment.

File: protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        // safe.setup(signers, 3, zero, "0x", 0xfd0732Dc9E303f09fCEf3a7388Ad10A83459Ec99, zero, 0, payable(zero));
+        // DEPLOYMENT USING SAFE PROXY
+
+        uint256 minDelay = 1 days;
+        address[] memory proposers = new address[](5);
+        address[] memory executors = new address[](1);
+
+        proposers[0] = vm.addr(5);
+        proposers[1] = vm.addr(6);
+        proposers[2] = vm.addr(7);
```
[the remark this answers is not in the document]
Comment (l-monninger) [reply protocol-units/settlement/mcr/contracts/script/DeployMoveTokenMultisig.s.sol:84]: Explain the lifetime of these signers, i.e., just during the script.

**Review by l-monninger — CHANGES_REQUESTED** [review]
Needs the following:
- Docs on intended usage.
- Assertions in the deployer script.
- Clean up of `run()` method.

**Review by l-monninger — APPROVED** [review]

**Pull request merged**